src/GovUk.Education.ExploreEducationStatistics.Content.Api/Services/ReleaseFileService.cs

```diff
@@ -48,31 +68,95 @@ public class ReleaseFileService : IReleaseFileService
                 .OnSuccess(async rf =>
                 {
                     return await GetBlob(rf.PublicPath())
-                        .OnSuccess(blob => DownloadToStream(blob, rf.File.Filename));
+                        .OnSuccess(blob => DownloadToStreamResult(blob, rf.File.Filename));
                 });
         }
 
-        public async Task<Either<ActionResult, FileStreamResult>> StreamByPath(string path)
+        public async Task<Either<ActionResult, Unit>> ZipFilesToStream(
+            Guid releaseId,
+            IList<Guid> fileIds,
+            Stream outputStream,
+            CancellationToken? cancellationToken = null)
         {
-            return await GetBlob(path)
-                .OnSuccess(DownloadToStream);
+            return await _persistenceHelper.CheckEntityExists<Release>(releaseId)
+                .OnSuccess(_userService.CheckCanViewRelease)
+                .OnSuccessVoid(
+                    async () =>
+                    {
+                        var releaseFiles = (await QueryByFileType(releaseId, ZipFileTypes)
+                            .Where(rf => fileIds.Contains(rf.FileId))
+                            .ToListAsync())
+                            .OrderBy(rf => rf.File.ZipFileEntryName())
+                            .ToList();
+
+                        await DoZipFilesToStream(releaseFiles, outputStream, cancellationToken);
+                    }
+                );
         }
 
-        private async Task<FileStreamResult> DownloadToStream(BlobInfo blob, string filename)
+        private async Task DoZipFilesToStream(
+            List<ReleaseFile> releaseFiles,
+            Stream outputStream,
+            CancellationToken? cancellationToken = null)
+        {
+            using var archive = new ZipArchive(outputStream, ZipArchiveMode.Create);
+
+            foreach (var releaseFile in releaseFiles)
+            {
+                // Stop immediately if we receive a cancellation request
+                if (cancellationToken?.IsCancellationRequested == true)
+                {
+                    return;
+                }
+
+                var blobExists = await _blobStorageService.CheckBlobExists(
+                    PublicReleaseFiles,
+                    releaseFile.PublicPath()
+                );
+
+                if (!blobExists)
+                {
+                    continue;
+                }
+
+                var entry = archive.CreateEntry(releaseFile.File.ZipFileEntryName());
+                await using var entryStream = entry.Open();
+
+                await _blobStorageService.DownloadToStream(
+                    containerName: PublicReleaseFiles,
+                    path: releaseFile.PublicPath(),
+                    stream: entryStream,
+                    cancellationToken: cancellationToken
+                );
+            }
+        }
+
+        public async Task<Either<ActionResult, FileStreamResult>> StreamAllFilesZip(Guid releaseId)
```

Comment:
Just thinking, will we still need the All Files zip when this functionality is eventually released? Not having it as a database File ref has caused us some overhead in the code that could be cleaned up if we can agree it's no longer required.

Reply:
Not fully fleshed out my plan for this - but I'm thinking about the following:

- Removing the 'All files' zip from the publishing phase completely
- Still have an endpoint to fetch the 'All files' zip. When requested for the first time, we zip this up in a similar manner to the `/files` endpoint.
- Once streamed to the user, we then upload the 'All files' zip to blob storage.
- On subsequent requests, we just re-fetch that blob, unless the release has been re-published, in which case we regenerate the blob (and delete the old one).

This would basically be a slight variation of our cache-aside strategy that we've done for the `BlobCache` stuff.

If this all sounds sensible to you, I'm planning on doing this all as part of this upcoming ticket: https://dfedigital.atlassian.net/browse/EES-2546

I think this approach has the advantage that we don't need to create a migration for existing 'All files' zips to add the data guidance text file to each one :smile:

Reply:
This sounds like a really good approach to cache the file and still only zip 'all' the files once. We get to clean stuff up and fix the problem with data guidance too. 🙂